HW_1/hw_1.py

```diff
@@ -0,0 +1,466 @@
+
+# Задание 1 (5 баллов)
+
+Напишите классы **Chat**, **Message** и **User**. Они должны соответствовать следующим требованиям:
+
+**Chat**:
++ Должен иметь атрибут `chat_history`, где будут храниться все сообщения (`Message`) в обратном хронологическом порядке (сначала новые, затем старые)
++ Должен иметь метод `show_last_message`, выводящий на экран информацию о последнем сообщении
++ Должен иметь метод `get_history_from_time_period`, который принимает два опциональных аргумента (даты с которой и по какую мы ищем сообщения и выдаём их). Метод также должен возвращать объект типа `Chat`
++ Должен иметь метод `show_chat`, выводящий на экран все сообщения (каждое сообщение в таком же виде как и `show_last_message`, но с разделителем между ними)
++ Должен иметь метод `recieve`, который будет принимать сообщение и добавлять его в чат
+
+**Message**:
++ Должен иметь три обязательных атрибута
+    + `text` - текст сообщения
+    + `datetime` - дата и время сообщения (встроенный модуль datetime вам в помощь). Важно! Это должна быть не дата создания сообщения, а дата его попадания в чат! 
+    + `user` - информация о пользователе, который оставил сообщение (какой тип данных использовать здесь, разберётесь сами)
++ Должен иметь метод `show`, который печатает или возвращает информацию о сообщении с необходимой информацией (дата, время, юзер, текст)
++ Должен иметь метод `send`, который будет отправлять сообщение в чат
+
+**User**:
++ Класс с информацией о юзере, наполнение для этого класса придумайте сами
+
+Напишите несколько примеров использования кода, которое показывает взаимодействие между объектами.
+
+В тексте задания намерено не указано, какие аргументы должны принимать методы, пускай вам в этом поможет здравый смысл)
+
+В этом задании не стоит флексить всякими продвинутыми штуками, для этого есть последующие
+
+В этом задании можно использовать только модуль `datetime`
+"""
+
+import datetime
+
+class Chat:
+    
+    def __init__(self, chat_history=[]):
+        self.chat_history = chat_history
+    
+    def show_last_message(self):
+        print(*self.chat_history[0])
+    
+    def get_history_from_time_period(self, date_from=None, date_to=None):
+        if date_from and date_to:
+            message_from = -1
+            message_to = 0
+            for message_ind in range(len(self.chat_history)-1, -1, -1):
+                if message_from == -1 and self.chat_history[message_ind][1] == date_from:
+                    message_from = message_ind
+                if self.chat_history[message_ind][1] == date_to:
+                    message_to = message_ind
+            return self.chat_history[message_to:message_from+1]
+        else:
+            return self.chat_history
+    
+    def show_chat(self):
+        for message in self.chat_history:
+            print(*message)
+    
+    def recieve(self, Message):
+        message = [Message.text, datetime.datetime.now().strftime("%d/%m/%y %H:%M"), Message.user.info()]
```

Comment:
1. Нейминг!!! Не надо называть переменные с большой буквы, особенно, когда у тебя в коде есть класс с таким именем
2. А почему бы не положить в `chat_history` сам объект `Message`? Зачем нам какой-то список, это же неудобно
3. Аналогично не совсем понятно, зачем превражать `datetime` в строку, это очень хороший объект, а ты берёшь и лишаешь его всего функционала

```suggestion
    def recieve(self, message):
        message = [message.text, datetime.datetime.now().strftime("%d/%m/%y %H:%M"), message.user.info()]
```